File: 53.maxSubArray.js
/**
 <p>Given an integer array <code>nums</code>, find the <span data-keyword="subarray-nonempty">subarray</span> with the largest sum, and return <em>its sum</em>.</p>

 <p>&nbsp;</p>
 <p><strong class="example">Example 1:</strong></p>

 <pre>
 <strong>Input:</strong> nums = [-2,1,-3,4,-1,2,1,-5,4]
 <strong>Output:</strong> 6
 <strong>Explanation:</strong> The subarray [4,-1,2,1] has the largest sum 6.
 </pre>

 <p><strong class="example">Example 2:</strong></p>

 <pre>
 <strong>Input:</strong> nums = [1]
 <strong>Output:</strong> 1
 <strong>Explanation:</strong> The subarray [1] has the largest sum 1.
 </pre>

 <p><strong class="example">Example 3:</strong></p>

 <pre>
 <strong>Input:</strong> nums = [5,4,-1,7,8]
 <strong>Output:</strong> 23
 <strong>Explanation:</strong> The subarray [5,4,-1,7,8] has the largest sum 23.
 </pre>

 <p>&nbsp;</p>
 <p><strong>Constraints:</strong></p>

 <ul>
 <li><code>1 &lt;= nums.length &lt;= 10<sup>5</sup></code></li>
 <li><code>-10<sup>4</sup> &lt;= nums[i] &lt;= 10<sup>4</sup></code></li>
 </ul>

 <p>&nbsp;</p>
 <p><strong>Follow up:</strong> If you have figured out the <code>O(n)</code> solution, try coding another solution using the <strong>divide and conquer</strong> approach, which is more subtle.</p>

 <div><div>Related Topics</div><div><li>Array</li><li>Divide and Conquer</li><li>Dynamic Programming</li></div></div><br><div><li>👍 33020</li><li>👎 1382</li></div>
 */

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * @param {number[]} nums
 * @return {number}
 */

var maxSubArray = function (nums) {
    // var keyword are hoisted and initialized which means they are accessible in their enclosing scope even before they are declared, however their value is undefined before the declaration statement is reached
    // let keyword are hoisted but not initialized which means they are not accessible in their enclosing scope before they are declared, however their value is undefined before the declaration statement is reached
    let result = -Infinity;
    let count = 0;
    for (let i = 0; i < nums.length; i++) {
        count += nums[i];
        if (count > result) {
            result = count;
        }
        // only reset count when the current sum is negative
        if (count < 0) {
            count = 0;
        }
    }
    return result;
};
//leetcode submit region end(Prohibit modification and deletion)
var nums = [-1];
console.log(maxSubArray(nums));

File: 53.maxSubArray.py
"""
Given an array of integers nums, find the subarray with the largest sum and return the sum.

A subarray is a contiguous non-empty sequence of elements within an array.

Example 1:

Input: nums = [2,-3,4,-2,2,1,-1,4]

Output: 8
Explanation: The subarray [4,-2,2,1,-1,4] has the largest sum 8.

Example 2:

Input: nums = [-1]

Output: -1
Constraints:

1 <= nums.length <= 1000
-1000 <= nums[i] <= 1000
14151613121117181097861920212223
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if len(nums) == 1: return nums[0]
        if len(nums) == 0: return 0
        # use dynamic programming to continue update the answer
        res = nums[0] # subarray is non-empty so at least take 1st value
        cur_sum = 0

        for i in range(0, len(nums)):

…
        return res 

--NORMAL--
Accepted

Passed test cases: 19 / 19
You have successfully completed this problem!
"""
# English description is not available for the problem. Please switch to Chinese
# . Related Topics 数组 分治 动态规划
#  👍 322 👎 0


# leetcode submit region begin(Prohibit modification and deletion)
# 输入: nums = [-2,1,-3,4,-1,2,1,-5,4]
# 输出: 6
# 解释: 连续子数组 [4,-1,2,1] 的和最大，为 6。
"""
# 提示：
#
# 1 <= arr.length <= 10^5
# -100 <= arr[i] <= 100
# 注意：本题与主站 53 题相同：https://leetcode-cn.com/problems/maximum-subarray/
"""
from typing import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if len(nums) == 1: return nums[0]
        if len(nums) == 0: return 0
        # use dynamic programming to continue update the answer
        res = nums[0] # subarray is non-empty so at least take 1st value
        cur_sum = 0

        for i in range(0, len(nums)):

            # only keep the positive result         
            if cur_sum < 0:
                cur_sum = nums[i]
            else:
                cur_sum += nums[i]
        

            # dp only record max for each sub array end position
            # need overall max
            res = max(res, cur_sum)

        return res



# leetcode submit region end(Prohibit modification and deletion)

s = Solution()
input_array = [2, -1, 1, 1]
maximum = s.maxSubArray(input_array)
print('maximum sum of sub array %s: %d' % ([2, -1, 1, 1], maximum))
